test_journal.py

```diff
@@ -0,0 +1,281 @@
+    # now, we should have one entry:
+    assert db_session.query(journal.Entry).count() == 1
+    for field in ['title', 'content']:
+        if field != 'session':
+            assert getattr(entry, field, '') == kwargs[field]
+    # id and created should be set automatically upon writing to db:
+    for auto in ['id', 'date']:
+        assert getattr(entry, auto, None) is not None
+
+
+def test_entry_no_title_fails(db_session):
+    bad_data = {'content': 'test text'}
+    journal.Entry.write(session=db_session, **bad_data)
+    with pytest.raises(IntegrityError):
+        db_session.flush()
+
+
+def test_entry_no_text_fails(db_session):
+    bad_data = {'title': 'test title'}
+    journal.Entry.write(session=db_session, **bad_data)
+    with pytest.raises(IntegrityError):
+        db_session.flush()
+
+
+def test_read_entries_empty(db_session):
+    entries = journal.Entry.all()
+    assert len(entries) == 0
+
+
+def test_read_entries_one(db_session):
+    title_template = "Title {}"
+    text_template = "Entry Text {}"
+    # write three entries, with order clear in the title and text
+    for x in range(3):
+        journal.Entry.write(
+            title=title_template.format(x),
+            content=text_template.format(x),
+            session=db_session)
+        db_session.flush()
+    entries = journal.Entry.all()
+    assert len(entries) == 3
+    assert entries[0].title > entries[1].title > entries[2].title
+    for entry in entries:
+        assert isinstance(entry, journal.Entry)
+
+
+@pytest.fixture()
+def app():
+    from journal import main
+    from webtest import TestApp
+    app = main()
+    return TestApp(app)
+
+
+def test_empty_listing(app):
+    response = app.get('/')
+    assert response.status_code == 200
+    actual = response.body
+    expected = 'No entries here so far'
+    assert expected in actual
+
+
+@pytest.fixture()
+def entry(db_session):
+    kwargs = {'title': "Test Title", 'content': "Test Entry Text"}
+    kwargs['session'] = db_session
+    # first, assert that there are no entries in the database:
+    # assert db_session.query(journal.Entry).count() == 0
+    # now, create an entry using the 'write' class method
+    entry = journal.Entry.write(**kwargs)
+    db_session.flush()
+    return entry
+
+
+def test_listing(app, entry):
+    response = app.get('/')
+    assert response.status_code == 200
+    actual = response.body
+    for field in ['title', 'content']:
+        expected = getattr(entry, field, 'absent')
+        assert expected in actual
+
+
+def test_post_to_add_view(app):
+    entry_data = {
+        'title': 'Hello there',
+        'content': 'This is a post',
+    }
+    response = app.post('/add', params=entry_data, status='3*')
+    redirected = response.follow()
+    actual = redirected.body
+    for expected in entry_data.values():
+        assert expected in actual
+
+
+def test_try_to_get(app):
+    with pytest.raises(webtest.AppError):
+        app.get('/add')
```

Comment:
hmmmm. Does `AppError` wrap something?  i would expect a 404 response from this request.